pkg/provisioner/ironic/ironic.go

```diff
@@ -1129,6 +1129,79 @@ func (p *ironicProvisioner) Adopt(force bool) (result provisioner.Result, err er
 	return
 }
 
+func (p *ironicProvisioner) startManualCleaning(ironicNode *nodes.Node) (result provisioner.Result, err error) {
+	var cleanSteps []nodes.CleanStep
+
+	// Start manual clean
+	if len(cleanSteps) != 0 {
+		p.log.Info("remove existing configuration and set new configuration", "steps", cleanSteps)
+		opts := nodes.ProvisionStateOpts{Target: nodes.TargetClean, CleanSteps: cleanSteps}
+		var success bool
+		success, result, err = p.tryChangeNodeProvisionState(ironicNode, opts)
+		if !success {
+			err = fmt.Errorf("start manual clean failed")
+		}
+	}
+	return
+}
+
+// Prepare remove existing configuration and set new configuration
+func (p *ironicProvisioner) Prepare() (result provisioner.Result, err error) {
+	var ironicNode *nodes.Node
+
+	if ironicNode, err = p.findExistingHost(); err != nil {
+		return result, errors.Wrap(err, "could not find host to clean")
+	}
+	if ironicNode == nil {
+		return result, fmt.Errorf("no ironic node for host")
+	}
+
+	switch nodes.ProvisionState(ironicNode.ProvisionState) {
+	case nodes.Available, nodes.Active:
+		// Do Nothing (go to provisioning state)
+
+	case nodes.Manageable:
+		if p.host.Status.ManualCleaning {
+			// Manual clean finished
+			p.host.Status.ManualCleaning = false
+			return
+		}
+		result, err = p.startManualCleaning(ironicNode)
+		if result.Dirty == true && err == nil {
+			p.host.Status.ManualCleaning = true
+		}
+		return
+
+	case nodes.CleanFail:
+		if p.host.Status.ManualCleaning {
+			p.host.Status.ManualCleaning = false
+			result.Dirty = true
+			return
+		}
+		if ironicNode.Maintenance {
+			p.log.Info("clearing maintenance flag")
+			return p.setMaintenanceFlag(ironicNode, false)
+		}
+		return p.changeNodeProvisionState(
+			ironicNode,
+			nodes.ProvisionStateOpts{Target: nodes.TargetManage},
+		)
+
+	default:
+		// Wait states like cleaning and clean wait
+		p.log.Info("waiting for host to become manageable",
+			"state", ironicNode.ProvisionState,
+			"deploy step", ironicNode.DeployStep)
+		// Avoid start manual cleaning success but set Cleaned flag failed
+		if !p.host.Status.ManualCleaning {
+			p.host.Status.ManualCleaning = true
+		}
```

Comment:
This one doesn't make me feel very comfortable, especially because is not performed in a restricted set of states (`cleaning` and `clean wait` as mentioned in the comment). If a cleaning in really in progress, IINW then it could be verified by checking that the `clean_steps` field of the `driver_internal_info` is not empty (see [1]).

[1] https://docs.openstack.org/ironic/latest/admin/cleaning.html

Reply:
Thanks, done.

Reply:
`DriverInternalInfo` it's a map that could contain other fields, so very likely that check must be modified.